daml/RealFac/Contract/Contract.daml

```
@@ -0,0 +1,162 @@
+module RealFac.Contract.Contract where
+
+import DA.Action
+import DA.List
+import RealFac.CashAccount.CashAccount
+import RealFac.Contract.ContractParties
+import RealFac.Contract.Phase
+
+template Contract
+  with
+    contractParties : ContractParties
+    ref : Text
+    phases : [Phase]
+    -- derived tacking data
+    done : Bool
+    toSettle : Int
+  where
+    signatory parties contractParties
+    observer parties contractParties
+    key (contractParties.accountant, ref) : (Party, Text)
+    maintainer key._1
+
+    let
+      ContractParties accountant buyer developer = contractParties
+
+    controller buyer can
+
+      Ack
+        : ContractId Contract
+        with
+          phaseId : Text
+        do
+          let
+            (newPhases, phase, allDone) = processPhase phaseId Done phases
+          create this with
+            phases = newPhases
+            done = allDone
+            toSettle = toSettle + phase.descriptor.cashOnComplete
+
+      Reject
+        : ContractId Contract
+        with
+          phaseId : Text
+        do
+          let
+            (newPhases, phase, allDone) = processPhase phaseId InProgress phases
+          create this with
+            phases = newPhases
+
+    controller developer can
+
+      Start
+        : ContractId Contract
+        with
+          phaseId : Text
+        do
+          let
+            (newPhases, phase, allDone) = processPhase phaseId InProgress phases
+          create this with
+            phases = newPhases
+
+      Complete
+        : ContractId Contract
+        with
+          phaseId : Text
+        do
+          let
+            (newPhases, phase, allDone) = processPhase phaseId PendingDone phases
+          create this with
+            phases = newPhases
+
+    controller accountant can
+
+      nonconsuming Settle
+        : ()
+        do
+          when (toSettle > 0) do
```

Is there a resaon to call `Settle` when `toSettle` is <= 0? If not, you could just make this a consuming choice and call `assert (toSettle > 0)` at the beginning. Then the transaction will fail once it’s <= 0 instead of creating empty transactions.

---

this is sort of a service choice which is called by automation (trigger), and I wanted this to be effect-less for the system if nothing need to be done, so the code is slicker on the triggerside